File: CS50git/cs50/pset2/crack/crack.c
// Take for input a hash (with salt)
// try to crack a 5-letter alphabetical password
#include <stdio.h>
#include <string.h>
#include <crypt.h>

int main(int argc, char *argv[])
{
    // Verify usage
    if (argc != 2)
    {
        printf("Usage: hash (with salt)\n");
        return 1;
    }

    // Init
    char *hash = argv[1];
    char salt[2];
    memcpy(salt, hash, 2);
    char password[6] = "\0\0\0\0\0\0";

    // Sorted through Robert Lewand's Cryptological Mathematics for faster crack
    char letters[53] = "\0etaoinshrdlcumwfgypbvkjxqzETAOINSHRDLCUMWFGYPBVKJXQZ";
    // 26 + 26 + 1 "\0" = 53
    const int letters_count = 53;


    // First try to match hash with 53 tries before adding another character
    // Then try 53^2 times before adding another character, etc...
    for (int fifthChar = 0; fifthChar < letters_count; fifthChar++)
    {
        for (int fourthChar = 0; fourthChar < letters_count; fourthChar++)
        {
            for (int thirdChar = 0; thirdChar < letters_count; thirdChar++)
            {
                for (int secondChar = 0; secondChar < letters_count; secondChar++)
                {
                    for (int firstChar = 1; firstChar < letters_count; firstChar++)
                    {
                        // Update password's characters
                        password[0] = letters[firstChar];
                        password[1] = letters[secondChar];
                        password[2] = letters[thirdChar];
                        password[3] = letters[fourthChar];
                        password[4] = letters[fifthChar];

                        // If hash matches, print password and return
                        if (!strcmp(crypt(password, salt), hash))
                        {
                            printf("%s\n", password);
                            return 0;
                        }
                    }
                }
            }
        }
    }
    return 2;
}

File: CS50git/cs50/pset6/more/crack.py
# Usage: python crack.py hash
# try up to 53^5 possibilities of encryption
# if success, return password
import sys
import crypt


def main(argv):

    # Verify usage
    if len(argv) != 2:
        print("Usage: python crack.py hash")
        exit(1)

    # Init
    USER_HASH = argv[1]
    SALT = USER_HASH[0:2]
    # Sorted through Robert Lewand's Cryptological Mathematics for faster crack
    LETTERS = " eEtTaAoOiInNsShHrRdDlLcCuUmMwWfFgGyYpPbBvVkKjJxXqQzZ"

    # Perform decryption
    for char5 in LETTERS:
        for char4 in LETTERS:
            for char3 in LETTERS:
                for char2 in LETTERS:
                    for char1 in LETTERS:

                        # Generate password
                        password = f"{char1}{char2}{char3}{char4}{char5}".strip()

                        # Try combination
                        if crypt.crypt(password, SALT) == USER_HASH:
                            print(password)
                            exit(0)

    # Couldn't be cracked
    exit(2)


if __name__ == "__main__":
    main(sys.argv)
